### PR TITLE
GC

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

**Comment by bryanmacfarlane** [comment]
We need a write up soon.  Please do as markdown in this as we've done with other docs recently.  We will RFC it.

File: src/Agent.Worker/Build/BuildDirectoryManager.cs [hunk 1]
```diff
@@ -124,6 +128,32 @@ public sealed class BuildDirectoryManager : AgentService, IBuildDirectoryManager
...
 
+        public void RunMaintenanceOperation(IExecutionContext executionContext)
+        {
+            Trace.Entering();
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            var trackingManager = HostContext.GetService<ITrackingManager>();
+            int staleBuildDirThreshold = executionContext.Variables.GetInt("maintenance.deleteworkingdirectory.daysthreshold") ?? 0;
+            if (staleBuildDirThreshold > 0)
+            {
+                staleBuildDirThreshold = Math.Min(staleBuildDirThreshold, 365);
```
Comment (ericsciple) [inline src/Agent.Worker/Build/BuildDirectoryManager.cs]: why min(___, 365)?

File: src/Agent.Worker/Build/TrackingManager.cs [hunk 2]
```diff
@@ -20,6 +23,10 @@ public interface ITrackingManager : IAgentService
         void MarkForGarbageCollection(IExecutionContext executionContext, TrackingConfigBase config);
 
         void UpdateJobRunProperties(IExecutionContext executionContext, TrackingConfig config, string file);
+
+        void DiscoverExpiredGarbage(IExecutionContext executionContext, TimeSpan expiration);
```
Comment (ericsciple) [inline src/Agent.Worker/Build/TrackingManager.cs]: "MarkExpiredForGarbageCollection"?

File: src/Agent.Worker/Build/TrackingManager.cs [hunk 3]
```diff
@@ -137,6 +144,112 @@ public void UpdateJobRunProperties(IExecutionContext executionContext, TrackingC
...
+                    // detect whether the tracking file is in new format.
+                    TrackingConfig newTracking = tracking as TrackingConfig;
+                    if (newTracking == null)
+                    {
+                        LegacyTrackingConfig legacyConfig = tracking as LegacyTrackingConfig;
+                        ArgUtil.NotNull(legacyConfig, nameof(LegacyTrackingConfig));
+
+                        Trace.Verbose($"{trackingFile} is a old format tracking file.");
+
+                        executionContext.Output($"Mark tracking file '{trackingFile}' for GC, since it never been used.");
```
Comment (ericsciple) [inline src/Agent.Worker/Build/TrackingManager.cs]: do all of these need to be loc'd?

Comment (ericsciple) [inline src/Agent.Worker/Build/TrackingManager.cs]: ..these strings written to executionContext.Output

File: src/Agent.Worker/Build/TrackingManager.cs [hunk 4]
```diff
@@ -137,6 +144,112 @@ public void UpdateJobRunProperties(IExecutionContext executionContext, TrackingC
...
+                        MarkForGarbageCollection(executionContext, legacyConfig);
+                        IOUtil.DeleteFile(trackingFile);
```
Comment (ericsciple) [inline src/Agent.Worker/Build/TrackingManager.cs]: i'm not sure how i feel about deleting all of the legacy tracking files :)

maybe delete is fine. although i'm wondering if we should fallback to timestamp on the file.

File: src/Agent.Worker/Build/TrackingManager.cs [hunk 5]
```diff
@@ -137,6 +144,112 @@ public void UpdateJobRunProperties(IExecutionContext executionContext, TrackingC
...
+                }
+            }
+        }
+
+        public void DisposeCollectedGarbage(IExecutionContext executionContext)
+        {
+            Trace.Entering();
+
+            string gcDirectory = Path.Combine(
+                IOUtil.GetWorkPath(HostContext),
```
Comment (ericsciple) [inline src/Agent.Worker/Build/TrackingManager.cs]: rebase?

File: src/Agent.Worker/Build/TrackingManager.cs [hunk 6]
```diff
@@ -137,6 +144,112 @@ public void UpdateJobRunProperties(IExecutionContext executionContext, TrackingC
...
+                {
+                    var gcConfig = LoadIfExists(executionContext, gcFile) as TrackingConfig;
+                    ArgUtil.NotNull(gcConfig, nameof(TrackingConfig));
+
+                    string fullPath = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), gcConfig.BuildDirectory);
+                    executionContext.Output($"Deleting '{fullPath}'");
+                    IOUtil.DeleteDirectory(fullPath, CancellationToken.None);
+
+                    executionContext.Output($"Delete gc tracking file after delete '{fullPath}'");
+                    IOUtil.DeleteFile(gcFile);
```
Comment (ericsciple) [inline src/Agent.Worker/Build/TrackingManager.cs]: it puts a smile on my face to see the gc files finally used for something :)

File: src/Agent.Worker/Maintenance/MaintenanceJobExtension.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    public sealed class MaintenanceJobExtension : AgentService, IJobExtension
+    {
+        public Type ExtensionType => typeof(IJobExtension);
+        public string HostType => "poolmaintenance";
+        public IStep PrepareStep { get; private set; }
+        public IStep FinallyStep { get; private set; }
+
+        public MaintenanceJobExtension()
+        {
+            PrepareStep = new JobExtensionRunner(
```
Comment (ericsciple) [inline src/Agent.Worker/Maintenance/MaintenanceJobExtension.cs]: Depending on which PR goes first, need to see what impact Condition has on this (references `Agent.JobStatus == succeeded`) for normal condition.

**Comment by ericsciple** [comment]
seems good

**Pull request merged**